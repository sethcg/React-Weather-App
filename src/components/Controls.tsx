'use client';

import { LatLng, Marker } from 'leaflet';
import { RefObject, useState } from 'react';

// import { WiUmbrella } from 'react-icons/wi';
import { FaMapMarkerAlt } from 'react-icons/fa';
import { CurrentWeather } from './WeatherType';

const iconClass =
  'bg-slate-500 hover:bg-slate-700 text-black text-lg font-extrabold p-2 rounded-xl';

interface Controls {
  markerRef: RefObject<Marker | null>;
}

export default function Controls({ markerRef }: Controls) {
  const lang = 'en';
  const units = 'imperial';

  const [weather, setWeather] = useState<CurrentWeather | null>(null);
  const [previousLatLng, setPreviousLatLng] = useState<LatLng | null>(null);

  const handleProcess = async () => {
    if (!markerRef.current) return;
    const latLng = markerRef.current.getLatLng();

    // Fetch only if the Marker LatLng has changed
    if (
      !previousLatLng ||
      (previousLatLng.lat !== latLng.lat && previousLatLng.lng !== latLng.lng)
    ) {
      // Fetch weather data
      await fetch(
        `https://api.openweathermap.org/data/2.5/weather?lat=${latLng.lat}&lon=${latLng.lng}&units=${units}&lang=${lang}&appid=${process.env.NEXT_PUBLIC_OPEN_WEATHER_API_KEY}`,
      )
        .then((response) => response.json())
        .then((json) => {
          // console.log(`[JSON]: ${JSON.stringify(json)}`);
          setWeather(json as CurrentWeather);
        })
        .then(() => {
          // Set previousLatLng to avoid duplicate fetching
          setPreviousLatLng(latLng);
        })
        .catch((error) => console.error(error));
    }
  };

  return (
    <>
      <div className="flex flex-row items-center justify-between gap-8 px-20">
        <button className={iconClass} onClick={handleProcess}>
          <FaMapMarkerAlt size={24} strokeWidth={2.5} />
        </button>
        <h1>Latitude: {weather?.coord.lat}</h1>
        <h1>Longitude: {weather?.coord.lon}</h1>
      </div>
    </>
  );
}
