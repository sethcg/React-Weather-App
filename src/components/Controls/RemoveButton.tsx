'use client'

import { FunctionComponent } from 'react'
import { IoClose } from 'react-icons/io5'

interface Props {
  handleRemove: () => Promise<void>
}

export const RemoveButton: FunctionComponent<Props> = ({ handleRemove }) => {
  const buttonClass = 'bg-red-800 hover:bg-red-600 text-black text-lg font-extrabold'

  return (
    <div className="flex w-full justify-end px-2">
      <button aria-label="remove" className={`flex cursor-pointer items-center rounded-2xl p-2 ${buttonClass}`} onClick={handleRemove}>
        <IoClose size={24} strokeWidth={2.5} />
      </button>
    </div>
  )
}
