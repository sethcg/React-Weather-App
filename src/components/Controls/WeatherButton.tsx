'use client'

import { FunctionComponent } from 'react'
import { Marker } from 'leaflet'
import { FaMapMarkerAlt } from 'react-icons/fa'
import { clsx } from 'clsx'

const roundCoordinate = (num: number): string => {
  return num.toFixed(3)
}

interface Props {
  handleProcess: () => Promise<void>
  mapMarker: Marker
}

export const WeatherButton: FunctionComponent<Props> = ({ handleProcess, mapMarker }) => {
  return (
    <div className="flex w-full flex-row items-center justify-between px-2">
      <button
        aria-label="weather"
        className={clsx(
          'w-full cursor-pointer rounded-xl p-2',
          'bg-emerald-700 text-neutral-950',
          'hover:bg-emerald-600 hover:drop-shadow-md hover:drop-shadow-emerald-500/50',
        )}
        onClick={handleProcess}>
        <div className="flex w-full flex-row items-center justify-between px-2">
          <FaMapMarkerAlt size={24} strokeWidth={2.5} />
          <div className="flex-grow text-xl font-bold">
            <span>Lat: {roundCoordinate(mapMarker.getLatLng().lat)}</span>
          </div>
          <div className="flex-grow text-xl font-bold">
            <span>Lon: {roundCoordinate(mapMarker.getLatLng().lng)}</span>
          </div>
        </div>
      </button>
    </div>
  )
}
