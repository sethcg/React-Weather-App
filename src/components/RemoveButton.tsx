'use client';

import { IoClose } from 'react-icons/io5';

const iconClass = 'bg-red-800 hover:bg-red-600 text-black text-lg font-extrabold';

interface Controls {
  handleRemove: () => Promise<void>;
}

export default function Controls({ handleRemove }: Controls) {
  return (
    <div className="flex w-full justify-end px-2">
      <button className={`flex items-center p-2 ${iconClass} rounded-2xl`} onClick={handleRemove}>
        <IoClose size={24} strokeWidth={2.5} />
      </button>
    </div>
  );
}
