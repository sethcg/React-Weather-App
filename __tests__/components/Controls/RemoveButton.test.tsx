import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import { RemoveButton } from '../../../src/components/Controls/RemoveButton'

describe('RemoveButton Snapshot', () => {
  it('Renders RemoveButton unchanged', async () => {
    const { container } = await waitFor(() => render(<RemoveButton handleRemove={async () => {}} />))

    expect(container).toMatchSnapshot()
  })
})

describe('RemoveButton', () => {
  it('RemoveButton on click', async () => {
    const mockCallBack = jest.fn()
    render(<RemoveButton handleRemove={mockCallBack} />)

    const button = await screen.findByRole('button')
    fireEvent.click(button)

    expect(mockCallBack).toHaveBeenCalled()
  })
})
