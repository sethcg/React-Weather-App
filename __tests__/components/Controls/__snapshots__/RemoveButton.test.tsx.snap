// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveButton Snapshot Renders RemoveButton unchanged 1`] = `
<div>
  <div
    class="flex w-full justify-end px-2"
  >
    <button
      aria-label="remove"
      class="flex cursor-pointer items-center rounded-2xl p-2 bg-red-800 hover:bg-red-600 text-black text-lg font-extrabold"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="2.5"
        viewBox="0 0 512 512"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m289.94 256 95-95A24 24 0 0 0 351 127l-95 95-95-95a24 24 0 0 0-34 34l95 95-95 95a24 24 0 1 0 34 34l95-95 95 95a24 24 0 0 0 34-34z"
        />
      </svg>
    </button>
  </div>
</div>
`;
